docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,260 @@
+    - [User Stories](#user-stories)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Node Prototype Pattern - Where we take a known good/working OS snapshot and make a “prototype” upon which all future node OS’ images are based.
+
+Warm nodes - The concept of keeping up extraneous and unused nodes to prevent having to wait for patches or security updates, thus allowing you to have those nodes ready when more users try to use your service without the wait times.
+
+Compute Image Gallery - A repository that is an Azure Resource for managing and sharing images
+
+Snapshot - A static copy of the image at the point in time that it is taken
+
+Prototype Node - Whichever node is chosen to clone the OS image of and cache it for future nodes to use
+
+## Summary
+
+The existing controllers will be modified to cache the Nodes’ OS image on a regular interval and update the model to use that image for future scale outs. This feature will:
+- allow for faster horizontal scaling 
+- prevent security alerts from new nodes which immediately need a security update as those new nodes will now come with the security update already installed
+- help prevent users from needing to spin up warm nodes and overprovision
+
+## Motivation
+
+A model scenario would be an operator spinning up a CAPZ cluster and having this feature be able to be toggled on or off. If it was toggled on, then as the months passed and more security updates and patches needed to be applied to the operator’s node OS image, these changes would be cached on a regular interval and the operator would no longer have to wait for these changes to apply on new node creations. Operators will also have the option of immediately propagating updates to nodes according to a RollingUpdate configuration or rely on the dynamic nature of the cluster to scale in updated nodes as needed by toggling a configuration. As a result, users will have faster horizontal scaling and require fewer warm nodes and overprovisioning to avoid this problem, especially since the new nodes will have the container images of the applications it will run pre-cached so pods will run quicker when scheduled. This feature will also help users have better security compliance as new nodes will already be compliant instead of needing to patch.
+
+### Goals
+
+1. A solution using the Node Prototype pattern which cache’s nodes’ OS image on a regular interval and replaces it as a new MachinePool template VM image
```

Comment:
this is the summary of the overall proposal 

goals might include: to make the feature opt-in (or configurable), to not break existing users, etc.

Think of what's "in scope" vs "out of scope" for this proposal

Reply:
Added those goals and reworded some of the others to make them sound more like what we generally want as end goals.